paimon-core/src/main/java/org/apache/paimon/mergetree/compact/PartialUpdateMergeFunction.java

```diff
@@ -190,11 +193,36 @@ private Factory(Options options, RowType rowType) {
         @Override
         public MergeFunction<KeyValue> create(@Nullable int[][] projection) {
             List<DataType> fieldTypes = tableTypes;
+            Map<Integer, SequenceGenerator> projFieldSequences = new HashMap<>();
+
             if (projection != null) {
                 fieldTypes = Projection.of(projection).project(tableTypes);
+                int[] fieldOldIndex = Projection.of(projection).toTopLevelIndexes();
+                Map<Integer, Integer> indexMap = new HashMap<>();
+                for (int i = 0; i < fieldOldIndex.length; i++) {
+                    indexMap.put(fieldOldIndex[i], i);
+                }
+                for (int index = 0; index < fieldOldIndex.length; index++) {
+                    int oldIndex = fieldOldIndex[index];
+                    if (fieldSequences.get(oldIndex) != null) {
+                        int oldKeyIndex = fieldSequences.get(oldIndex).index();
+                        int keyIndex = indexMap.get(oldKeyIndex);
+                        if (!projFieldSequences.containsKey(keyIndex)) {
+                            SequenceGenerator newSequenceGen =
+                                    new SequenceGenerator(keyIndex, fieldTypes.get(keyIndex));
```

Comment:
keyIndex is not new index of oldIndex.
There should be a bug! You should add tests to cover.